### DISCUSSION TITLE
Sending a list with the Pyboard library

### DISCUSSION THREAD

**Discussion by roaldarbol** [micropython/micropython#9495] [Using MicroPython]
This is actually a continuation of an issue on the MicroPython forum: https://forum.micropython.org/viewtopic.php?f=2&t=12663. @dhylands perhaps you can be of help here again?

Normally I would just use `serial.readline()`, however I'm using Pyboard from Python and that doesn't seem to work (I don't think it's supported?). I can easily send individual bytes, however I can't find a way to send whole lines? My current issue is that I want to create an object on the computer that contains the files on the board. I can use `os.listdir()` and also print it well formatted - but I have not been able to find a way to re-create the list. 
The list consists of 2 files: `test.py` and `talk.py`.

On my Pico I have `talk.py`:
```python
import sys
import select
import os

poll = select.poll()
poll.register(sys.stdin, select.POLLIN)

def print_list():
    print(os.listdir())
```

And locally I run:
```python
import select
import time
from mpremote import pyboard
from serial.tools import list_ports

def log_for(secs):
    global pyb
    info = {}
    end = time.time() + secs
    while time.time() < end:
        events = poll.poll(1000)
        for file in events:
              if file[0] == pyb.serial.fileno():
                  ch = pyb.serial.read(1) # readline() never stops 
                  print(ch.decode('utf-8'), end='')

# Initialize PyBoard and connect in REPL mode
port = list(list_ports.comports())
pyb = pyboard.Pyboard(port[1].device, 115200)

poll = select.poll()
poll.register(pyb.serial, select.POLLIN)

pyb.enter_raw_repl()
pyb.execfile('talk.py')
pyb.exec_raw_no_follow('print_list()')
log_for(1)
```

And I get this relatively neat printed output:
```
['test.py', 'talk.py']
  >
```

How do I get this as a list I can work with?

**Comment by roaldarbol** [comment]
A related problem for other uses, I'd like to log the output with the `logging` library - here my issue is quite similar, I can't emulate the `print(ch.decode('utf-8'), end='')` call.

**Comment by robert-hh** [comment]
Instead of pyb.serial you can use the machine.UART class, which supports readline. machine.UART is portable across all ports. The pyb.xx classes are specific to PyBoard and kept for legacy, but new code should used the machine classes.

Edit: pyb.UART supports readline as well, so I'm wondering which board and firmware you use?
Edit2: So you do not use the PyBoard board, but the pyboard.py software with a PC.

**Comment by jimmo** [comment]
> And I get this relatively neat printed output:
> 
> ```
> ['test.py', 'talk.py']
>   >
> ```
> 
> How do I get this as a list I can work with?

If you trust both sides of this, then `eval(pyb.eval("expression"))` is a really easy option here, especially used in conjunction with `repl` to get it in a format that can be eval'ed.

If you `print(repr(something))` on the board, then `eval(pyb.exec_raw_no_follow("..."))` will do what you want.

Even simpler:

`files = eval(pyb.eval(f"os.listdir('{path}')"))`

**Reply by jimmo** [reply]
But don't use this if you don't trust the code running on the board!

**Comment by jimmo** [comment]
Another great option for this use case is https://github.com/BrianPugh/belay

**Comment by TheChymera** [comment]
@jimmo 

So I tried the suggestions from this discussion and:

```python
from mpremote import pyboard
from serial.tools import list_ports

# Initialize PyBoard and connect in REPL mode
port = list(list_ports.comports())
pyb = pyboard.Pyboard(port[0].device, 115200)

mypath="."

files = eval(pyb.eval(f"import os; os.listdir('{mypath}')"))

print([f for f in files])
```

doesn't print anything. It just hangs indefinitely. Any ideas what I'm doing wrong?

**Reply by jimmo** [reply]
> doesn't print anything. It just hangs indefinitely. Any ideas what I'm doing wrong?

My guess is that the first serial port isn't the one your device is actually connected to.
